test/pack/pack-common.c

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <assert.h>
+#include <string.h>
+#include <pthread.h>
+#include "yaksa_config.h"
+#include "yaksa.h"
+#include "dtpools.h"
+#include "pack-common.h"
+
+void pack_init_devices(void)
+{
+#ifdef HAVE_CUDA
+    pack_cuda_init_devices();
+#elif defined(HAVE_ZE)
+    pack_ze_init_devices();
+#endif
+}
+
+void pack_finalize_devices()
+{
+#ifdef HAVE_CUDA
+    pack_cuda_finalize_devices();
+#elif defined(HAVE_ZE)
+    pack_ze_finalize_devices();
+#endif
+}
+
+void pack_alloc_mem(size_t size, mem_type_e type, void **hostbuf, void **devicebuf)
+{
+    if (type == MEM_TYPE__UNREGISTERED_HOST) {
+        *devicebuf = malloc(size);
+        if (hostbuf)
+            *hostbuf = *devicebuf;
+#ifdef HAVE_CUDA
+    } else {
+        pack_cuda_alloc_mem(size, type, hostbuf, devicebuf);
+    }
+#elif defined(HAVE_ZE)
+    } else {
+        pack_ze_alloc_mem(size, type, hostbuf, devicebuf);
+    }
+#else
+    } else {
+        fprintf(stderr, "ERROR: no GPU device is supported\n");
+        exit(1);
+    }
+#endif
```

Comment:
This will not compile if yaksa finds both CUDA and ZE libraries on the system.  How about the following style instead:

```
if (...) {
   /* some stuff */
#Ifdef defined(HAVE_CUDA)
} else {
   /* some stuff */
#if defined(HAVE_ZE)
} else {
   /* some stuff */
#endif
} else {
   /* some stuff */
}
```

Reply:
This does not work if either HAVE_CUDA and HAVE_ZE is defined, because it will have two "else" branches.

Reply:
Right.  Something like this might work.  If not, some experimentation can help you figure out what will work:

```
if (...) {
   /* some stuff */
} else {
#ifdef defined(HAVE_CUDA)
   /* some stuff */
#elif defined(HAVE_ZE)
   /* some stuff */
#else
   /* some stuff */
#endif
}
```

Reply:
changed